### PR TITLE
E2E: Add plugin-e2e scenario verification tests

### PR DESCRIPTION
**Pull request by sunker**
**What is this feature?**

Lately, Plugins platform have been working on a [new library](https://github.com/grafana/plugin-tools/tree/main/packages/plugin-e2e) that adds improved E2E testing capabilities to Grafana plugins (more information in [this](https://github.com/grafana/grafana/issues/77484) issue and in [this](https://docs.google.com/document/d/1vLId8Gv8OQWqZZbgHfpy91ljARzs5ihfN7KxEQdN07U/edit?usp=sharing) design doc). The new library will (eventually) supersede the [`grafana/e2e`](https://github.com/grafana/grafana/blob/main/packages/grafana-e2e/) testing framework which haven't been actively maintained in a long time. The new package is called [`grafana/plugin-e2e`](https://www.npmjs.com/package/@grafana/plugin-e2e) and unlike `grafana/e2e`, `grafana/plugin-e2e` is an external npm package that is completely detached from the Grafana release cycle. 

`grafana/plugin-e2e` is based on [Playwright test](https://playwright.dev/). However, plugin-e2e does not bundle Playwright. This means the consumer (Grafana in this case) needs to install it, hence there are two new dev dependencies added to the root workspace. To run Playwright tests in one or many browsers, the browser binaries need to be installed on the machine. In this PR, tests are only run in Chromium. Installing browser binaries using the Playwright cli is straight forward, and I've added instructions on how to do this in the developer guide. 

Playwright tests will be executed in new Drone step in CI. All tests run in parallel and it takes roughly 1 minute if all tests pass. Playwright will generate an html report for every session, and the report is pushed to GCS in a subsequent step. You can browse the report generated for this PR [here](https://storage.googleapis.com/releng-pipeline-artifacts-dev/153215/playwright-report/index.html). 

**Why do we need this feature?**

We want to bring E2E testing capabilities to core plugins (and core plugins that run as external plugins). Second, and most importantly, we want to verify that the fixtures, models and expect matchers that are exported by `grafana/plugin-e2e` are compatible with the main branch of Grafana. 

If for example the UI for navigating from a dashboard page to a new panel changes in Grafana, it may break the code that navigates to the panel edit page in `grafana/plugin-e2e`. The panel edit page scenario exposed by `grafana/plugin-e2e` may be used in hundreds of E2E tests in dozens of plugins, so we'd want to prevent that from happening by capturing this before the change in Grafana is merged to main. This PR adds a bunch of E2E tests that has the purpose of verifying that the scenarios exported by grafana/plugin-e2e are still working as expected. 

In case a code change in Grafana core breaks a scenario in `grafana/plugin-e2e`, it will be the contributor's responsibility to come up with the necessary fix in `grafana/plugin-e2e`. If a scenario verification test fails, it will display a link to the [contributing guidelines](https://github.com/grafana/plugin-tools/blob/976bd5d3edd77d4864ec37ce18e5cb181c733ce9/packages/plugin-e2e/CONTRIBUTING.md#how-to-fix-broken-test-scenarios-after-changes-in-grafana) in the plugin-e2e repo where they can find detailed instructions on how fix, test and deploy their feature.

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

Plugin authors

**Which issue(s) does this PR fix?**:

Part of https://github.com/grafana/grafana/issues/77484

Fixes #

**Special notes for your reviewer:**

It's very important that the scenario verification tests added in this PR are stable and don't generate false negatives.For that reason, I'm not preventing contributors to merge PRs in the case Playwright tests fail. I'll change that when these tests have been running in main for some time. 

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: plugin-e2e/verify-scenario/README.md
Comment (sunker) [inline plugin-e2e/verify-scenario/README.md:8]: If the link 404s it's because [this](https://github.com/grafana/plugin-tools/pull/634) PR is not merged in the plugin-tools repo yet.

**Comment by aangelisc** [comment]
This looks really great to me @sunker! The only immediate question I have is if the bucket should be publically viewable or not 🤔

**Comment by sunker** [comment]
> The only immediate question I have is if the bucket should be publically viewable or not 🤔

It's a good question @aangelisc! I _think_ it should be fine since the we don't provision the Grafana instance we use for the E2E tests with any sensitive information. Also we have been pushing cypress video recordings to the same bucket for a long time, and these are also publicly available. For external plugins though, we should be more cautious. WDYT?

**Review by baldm0mma — APPROVED** [review]
This looks great! Any timeline ideas on when this will fully supersede/replace `grafana/e2e`?

File: plugin-e2e/verify-scenario/panelEditPage.spec.ts
Comment (mckn) [inline plugin-e2e/verify-scenario/panelEditPage.spec.ts:38]: Question: How come we don't have a function directly on the panelEditPage instead of using the datasource property? I'm asking just from a consistency perspective since we have `set` or `get` for most other things directly on the page.

```ts
await panelEditPage.setDatasource('gdev-testdata');
```

File: scripts/drone/steps/lib.star
Comment (mckn) [inline scripts/drone/steps/lib.star:376]: Will these steps run on every PR? Is there a risk of someone else opening a PR where they change the build script and just echo the secrets in the console so it get exposed?

**Review by mckn — COMMENTED** [review]
Wow, this looks great! Love the tests and the structure of this. Amazing work @sunker 🙌🏻  There is one thing regarding the build scripts that we probably need to have a look at but other than that it looks great.

File: plugin-e2e/verify-scenario/panelEditPage.spec.ts
Comment (sunker) [reply plugin-e2e/verify-scenario/panelEditPage.spec.ts:38]: I've been going for the composable approach here - the DataSourcePicker is a model which is being used across a bunch of different pages. I should be consistent in the sense that the data source is set the same way in the variable edit page, annoation edit page, explore etc.

File: scripts/drone/steps/lib.star
Comment (sunker) [reply scripts/drone/steps/lib.star:376]: As discussed offline, the Drone build does not run if the changes in drone.yaml aren't signed, and only org members have access to the key used for signing.

File: playwright.config.ts
Comment (aangelisc) [inline playwright.config.ts:17]: Why do we want to opt out of parallel tests on CI? Is this unsupported?

**Review by aangelisc — COMMENTED** [review]
This is looking really good to me @sunker! Thank you for building this out. 

The only key scenario I can think of that should be added would be for alerting (creation of an alert rule). 

Additionally, I wonder if the code-owners for the individual specs should be @grafana/plugins-platform-frontend or the squad responsible for that specific area of Grafana. Wdyt?

**Comment by sunker** [comment]
> The only key scenario I can think of that should be added would be for alerting (creation of an alert rule).

Yeah I think this would be useful, and definitely something on our todolist. But I don't think it's completely necessary to have in place when we release 1.0.0 of plugin-e2e.   

> Additionally, I wonder if the code-owners for the individual specs should be @grafana/plugins-platform-frontend or the squad responsible for that specific area of Grafana. Wdyt?

I think that would make sense. But I would like to see these tests run in main for some time first. Just want to make sure they run without flakiness before we start handing over responsibility to other teams. Does that sound good? 

I'm aiming to merge this PR something this week - just waiting for one more feature to merged in plugin-e2e first.

File: playwright.config.ts
Comment (sunker) [reply playwright.config.ts:17]: Following Playwright [standards](https://playwright.dev/docs/ci#workers) here. We may be able to fine-tune this in the future when we have more tests, but let's start simple. The tests only takes a few seconds to run anyway.

**Comment by baldm0mma** [comment]
Love all the added tests! Looks like there's a few conflicts, but once resolved, still LGTM. :)

File: .drone.yml
Comment (mckn) [inline .drone.yml:657]: Q: added to resolve some kind of timing issues?

File: .drone.yml
Comment (mckn) [inline .drone.yml:665]: Q: shouldn't we fail the pipeline if this step fails?

File: e2e/plugin-e2e/plugin-e2e-api-tests/README.md
Comment (mckn) [inline e2e/plugin-e2e/plugin-e2e-api-tests/README.md:1]: I like this documentation with super clear instructions on how to address potential failure! Great work!

File: e2e/plugin-e2e/plugin-e2e-api-tests/as-admin-user/explorePage.spec.ts
Comment (mckn) [inline e2e/plugin-e2e/plugin-e2e-api-tests/as-admin-user/explorePage.spec.ts:6]: Am I reading this wrong or is page unused?

```suggestion
test('query data response should be OK when query is valid', async ({ explorePage }) => {
  await explorePage.datasource.set('gdev-testdata');
```

File: e2e/plugin-e2e/plugin-e2e-api-tests/as-admin-user/featureToggles.spec.ts
Comment (mckn) [inline e2e/plugin-e2e/plugin-e2e-api-tests/as-admin-user/featureToggles.spec.ts:1]: Love these APIs! Super easy to work with 🙌🏻

**Review by mckn — APPROVED** [review]
Great work on this! LGTM! ⭐

File: .drone.yml
Comment (sunker) [reply .drone.yml:665]: Good catch - should have made this clear in the comment above...yes we should eventually! Just want to see this step run in the main branch for a week or so first. Want to be certain that these tests don't introduce any kind of flakiness.

File: .drone.yml
Comment (sunker) [reply .drone.yml:657]: Yes! It seems sometimes the server is not ready when playwright tests starts, so adding this just to be sure. Test tests run fast, so this shouldn't have an impact on the overall build time. 

Note that this file is generated from the lib.start file. I have added a comment [here](https://github.com/grafana/grafana/blob/0285aef332aeee7d9e505487a5dc833c82175402/scripts/drone/steps/lib.star#L862), but comments are being stripped out during file generation.

File: .drone.yml
Comment (mckn) [reply .drone.yml:665]: Make sense 👍🏻

**Pull request merged**
